src/cmd/factory/impl/common.go

```diff
@@ -179,3 +199,547 @@ func buildCollections(
 
 	return collections, nil
 }
+
+type permData struct {
+	user        string // user is only for older versions
+	entityID    string
+	roles       []string
+	sharingMode onedrive.SharingMode
+}
+
+type itemData struct {
+	name  string
+	data  []byte
+	perms permData
+}
+
+type itemInfo struct {
+	// lookupKey is a string that can be used to find this data from a set of
+	// other data in the same collection. This key should be something that will
+	// be the same before and after restoring the item in M365 and may not be
+	// the M365 ID. When restoring items out of place, the item is assigned a
+	// new ID making it unsuitable for a lookup key.
+	lookupKey string
+	name      string
+	data      []byte
+}
+
+type onedriveCollection struct {
+	service       path.ServiceType
+	pathElements  []string
+	items         []itemInfo
+	aux           []itemInfo
+	backupVersion int
+}
+
+type onedriveColInfo struct {
+	pathElements []string
+	perms        permData
+	files        []itemData
+	folders      []itemData
+}
+
+var (
+	folderAName = "folder-a"
+	folderBName = "b"
+	folderCName = "folder-c"
+
+	fileAData = []byte(strings.Repeat("a", 33))
+	fileBData = []byte(strings.Repeat("b", 65))
+	fileEData = []byte(strings.Repeat("e", 257))
+
+	// Cannot restore owner or empty permissions and so not testing them
+	writePerm = []string{"write"}
+	readPerm  = []string{"read"}
+)
+
+func generateAndRestoreOnedriveItems(
+	gc *connector.GraphConnector,
+	resourceOwner, secondaryUserID, secondaryUserName string,
+	acct account.Account,
+	service path.ServiceType,
+	cat path.CategoryType,
+	sel selectors.Selector,
+	tenantID, destFldr string,
+	count int,
+	errs *fault.Bus,
+) (
+	*details.Details,
+	error,
+) {
+	ctx, flush := tester.NewContext()
+	defer flush()
+
+	// TODO: fit the destination to the containers
+	dest := control.DefaultRestoreDestination(common.SimpleTimeTesting)
+	dest.ContainerName = destFldr
+	print.Infof(ctx, "Restoring to folder %s", dest.ContainerName)
+
+	d, _ := gc.Service.Client().UsersById(resourceOwner).Drive().Get(ctx, nil)
+	driveID := ptr.Val(d.GetId())
+
+	var (
+		cols []onedriveColInfo
+
+		rootPath = []string{
+			"drives",
+			driveID,
+			"root:",
+		}
+		folderAPath = []string{
+			"drives",
+			driveID,
+			"root:",
+			folderAName,
+		}
+		folderBPath = []string{
+			"drives",
+			driveID,
+			"root:",
+			folderBName,
+		}
+
+		folderCPath = []string{
+			"drives",
+			driveID,
+			"root:",
+			folderCName,
+		}
+
+		now         = time.Now()
+		currentTime = fmt.Sprintf("%d-%d-%d", now.Hour(), now.Minute(), now.Second())
+	)
+
+	for i := 0; i < count; i++ {
+		var col onedriveColInfo
+
+		// basic folder and file creation
+		if i == 0 {
```

Comment:
This pattern (including the modulus below) is more clever than helpful.  A simpler pattern would be to say ("for each count in --count, the factory will construct a set of files that adequately express permission").  IE: if `--count 1`, then you create 4 files (I think that's the standard, right?).  if `--count 2`, you repeat the set to make 8 files, and so on.  Additional fiddling with complexity beyond that isn't of any benefit.